src/cpp/shared/artifacts.cpp

```diff
@@ -15,437 +19,72 @@
-	"A dirty shovel has been thrust into a dirt mound nearby. Upon investigation, you discover it to be the enchanted shovel of the Gravediggers, long thought lost by mortals.",
-	"You pass by a chest lying under a tree. Suddenly, a goblin jumps out of the chest and shrieks. Before it can raise its club, you grab it and throw it back in the chest, and tie it sealed with rope.",
-    "You get a sudden urge to dig where you are standing. You dig and find a glistening metal gauntlet, mysteriously untouched by dirt. As you touch it, it leaps up and fuses itself onto your arm."
-};
+  template <typename T>
+  std::string JoinSequence(const T &seq) {
+    std::string ret;
+    for (const auto &str : seq) {
+      ret += str;
+    }
+    return ret;
+  }
+}
 
-unsigned char gArtifactLevel[] =
-{
-  254,
-  1,
-  1,
-  1,
-  1,
-  1,
-  1,
-  1,
-  2,
-  4,
-  4,
-  2,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  2,
-  4,
-  2,
-  4,
-  2,
-  4,
-  2,
-  2,
-  2,
-  2,
-  2,
-  4,
-  2,
-  2,
-  8,
-  8,
-  8,
-  8,
-  2,
-  4,
-  4,
-  8,
-  4,
-  4,
-  8,
-  8,
-  8,
-  4,
-  4,
-  4,
-  8,
-  4,
-  4,
-  4,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  8,
-  2,
-  8,
-  8,
-  8,
-  2,
-  4,
-  4,
-  8,
-  8,
-  2,
-  2,
-  2,
-  4,
-  4,
-  2,
-  4,
-  2,
-  4,
-  2,
-  16,
-  32,
-  32,
-  32,
-  32,
-  2,
-  8,
-  2,
-  8,
-  2,
-  2,
-  8,
-  8,
-  2,
-  2,
-  2,
-  4,
-  2,
-  2,
-  2,
-  2,
-  4,
-  4,
-  2
-};
\ No newline at end of file
+
+char *gArtifactNames[NUM_SUPPORTED_ARTIFACTS] = { 0 };
+char *gArtifactDesc[NUM_SUPPORTED_ARTIFACTS] = { 0 };
+char *gArtifactEvents[NUM_SUPPORTED_ARTIFACTS] = { 0 };
+unsigned char gArtifactLevel[NUM_SUPPORTED_ARTIFACTS] = { 0 };
+
+std::vector<std::string> names;
+std::vector<std::string> descriptions;
+std::vector<std::string> events;
+
+
+void LoadArtifacts() {
+  auto allArtifacts = artifacts_("./DATA/artifacts.xml");
+  const auto &artifactList = allArtifacts->artifact();
+
+  const int numArtifacts = TotalArtifacts(artifactList);
+  names.resize(numArtifacts);
+  descriptions.resize(numArtifacts);
+  events.resize(numArtifacts);
+
+  for (const auto &art : artifactList) {
+    const int i = art.id();
+    if (i < 0 || i >= numArtifacts) {
+      continue;
+    }
+
+    names[i] = art.name();
+    gArtifactNames[i] = &(names[i][0]);
+
+    if (art.description().size() > 0 && art.title().size() > 0) {
+      std::string desc = "{";
+      desc += JoinSequence(art.title());
+      desc += "}\n\n";
+      desc += JoinSequence(art.description());
+      descriptions[i] = desc;
+    }
+    gArtifactDesc[i] = &(descriptions[i][0]);
+
+    events[i] = JoinSequence(art.event());
+    gArtifactEvents[i] = &(events[i][0]);
+
+    const auto lvl = static_cast<unsigned char>(std::pow(2, static_cast<int>(art.level())));
```

Comment:
I don't see how this results in the correct value being assigned.

Reply:
The level in the xml file is an enum type.  Ultimate artifacts have level 0, major level 1, minor level 2, and treasure level 3.  The game expects numbers of 1, 2, 4, and 8 respectively.  I'll add a comment explaining this better.

Reply:
Okay, so it's assigned in order of appearance in the XSD file. Is there any way to make this more clear in the  So, you're saying that 32 is the lowest level of artifact rather than the highest one?

I'm more comfortable with an explicit mapping from level name to integer, decoupling us from the order in the XSD file and making this more explicit.